### PR TITLE
Cleanup

### PR DESCRIPTION
**Pull request by Clyybber**

### CODE REVIEW

File: compiler/lexer.nim [hunk 1]
```diff
@@ -962,15 +994,13 @@ proc getPrecedence*(tok: TToken, strongSpaces: bool): int =
     of '?': result = 2
     else: considerAsgn(2)
   of tkDiv, tkMod, tkShl, tkShr: result = 9
-  of tkIn, tkNotin, tkIs, tkIsnot, tkOf, tkAs, tkFrom: result = 5
   of tkDotDot: result = 6
+  of tkIn, tkNotin, tkIs, tkIsnot, tkOf, tkAs, tkFrom: result = 5
   of tkAnd: result = 4
   of tkOr, tkXor, tkPtr, tkRef: result = 3
   else: return -10
-  result = considerStrongSpaces(result)
```
Comment (timotheecour) [inline compiler/lexer.nim:970]: is that a behavior change?

File: compiler/lexer.nim [hunk 2]
```diff
@@ -1173,16 +1203,14 @@ proc skip(L: var TLexer, tok: var TToken) =
...
+proc rawGetTok*(L: var Lexer, tok: var Token) =
   template atTokenEnd() {.dirty.} =
     when defined(nimsuggest):
       # we attach the cursor to the last *strong* token
       if tok.tokType notin weakTokens:
         L.previousToken.line = tok.line.uint16
         L.previousToken.col = tok.col.int16
 
-  when defined(nimsuggest):
-    L.cursor = CursorPosition.None
```
Comment (timotheecour) [inline compiler/lexer.nim:1185]: why?
likewise with other places where cursor was removed (yes i know about the cursor comment, but given size of this PR, maybe it's best to remove cursor in another PR)

File: compiler/lineinfos.nim [hunk 3]
```diff
@@ -27,135 +27,91 @@ proc createDocLink*(urlSuffix: string): string =
...
+    errInternal = "internal error: $1"
+    errIllFormedAstX = "illformed AST: $1"
+    errCannotOpenFile = "cannot open '$1'"
+    errXExpected = "'$1' expected"
+    errGridTableNotImplemented = "grid table is not implemented"
+    errGeneralParseError = "general parse error"
+    errNewSectionExpected = "new section expected"
+    errInvalidDirectiveX = "invalid directive: '$1'"
+    errProveInit = "Cannot prove that '$1' is initialized."
+    # deadcode
```
Comment (timotheecour) [inline compiler/lineinfos.nim]: indent `deadcode` so it keeps referring to `errProveInit`

File: compiler/lexer.nim (hunk 1 above)
Comment (Clyybber) [reply compiler/lexer.nim:970]: No, the strongSpaces parsing mode is long gone.

File: compiler/lexer.nim (hunk 2 above)
Comment (Clyybber) [reply compiler/lexer.nim:1185]: It's dead/noop code, thats why I removed it.

File: compiler/lineinfos.nim (hunk 3 above)
Comment (Clyybber) [reply compiler/lineinfos.nim]: Good catch!

File: compiler/renderer.nim [hunk 4]
```diff
@@ -240,28 +240,6 @@ proc maxLineLength(s: string): int =
       inc(lineLen)
       inc(i)
 
-proc putRawStr(g: var TSrcGen, kind: TTokType, s: string) =
```
Comment (timotheecour) [inline compiler/renderer.nim:253]: deadcode but wondering if can be useful (with a deadcode comment) or safe to delete

File: compiler/wordrecg.nim [hunk 5]
```diff
@@ -19,74 +19,302 @@ from strutils import cmpIgnoreStyle
...
+    wImportCpp = "importcpp"
+    wImportObjC = "importobjc"
+    wImportCompilerProc = "importCompilerProc"
+    wImportc = "importc"
+    wImportJs = "importjs"
+    wExportc = "exportc"
+    wExportCpp = "exportcpp"
+    wExportNims = "exportnims"
+    wIncompleteStruct = "incompleteStruct"
+    # deprecated
```
Comment (timotheecour) [inline compiler/wordrecg.nim]: ditto, indent (but can be done in separate cleanup PR if needed)

File: compiler/parser.nim [hunk 6]
```diff
@@ -800,25 +798,25 @@ proc primarySuffix(p: var TParser, r: PNode,
       # `foo ref` or `foo ptr`. Unfortunately, these two are also
       # used as infix operators for the memory regions feature and
       # the current parsing rules don't play well here.
-      if p.inPragma == 0 and (isUnary(p) or p.tok.tokType notin {tkOpr, tkDotDot}):
+      if p.inPragma == 0 and (isUnary(p.tok) or p.tok.tokType notin {tkOpr, tkDotDot}):
```
Comment (timotheecour) [inline compiler/parser.nim:801]: isUnary(p)
vs
isUnary(p.tok)
?

Comment (timotheecour) [reply compiler/parser.nim:801]: nevermind, saw you also changed proc to `proc isUnary(tok: Token)`

File: compiler/vmgen.nim [hunk 7]
```diff
@@ -1360,19 +1360,6 @@ proc genMarshalStore(c: PCtx, n: PNode, dest: var TDest) =
   c.gABx(n, opcMarshalStore, 0, c.genType(n[1].typ))
   c.freeTemp(tmp)
 
-const
```
Comment (timotheecour) [inline compiler/vmgen.nim:1363]: deadcode (to be marked as such with a comment) by maybe useful?

File: compiler/lexer.nim (hunk 2 above)
Comment (timotheecour) [reply compiler/lexer.nim:1185]: indeed, i verified cursor is set but never read, ok!

**Comment by timotheecour** [comment]
/cc @Araq I've reviewed everything, looks excellent! feel free to review but I've already done a thorough pass and LGTM (modulo remaining comments) so feel free to skip this one if you want to save time.

@Clyybber 
* just 1 more question:
TSpecialWord elements have 2 stringifications:
via `canonPragmaSpelling` and via `specialWords`; in your PR, `canonPragmaSpelling` overrides `specialWords` when `canonPragmaSpelling` differs; was there any need for the non-canonical spellings (eg `executeonreload` vs `executeOnReload` for `wExecuteOnReload`)

* if possible, add a small summary in PR top message for posterity where you list just a few salient points addressed in this PR (eg: removed lexer.cursor, etc)

* did you use any special tricks to find deadcode?

File: compiler/wordrecg.nim (hunk 5 above)
Comment (Clyybber) [reply compiler/wordrecg.nim]: Thanks! If I already do a big cleanup I might as well cleanup the junk that I leave around :D

File: compiler/renderer.nim (hunk 4 above)
Comment (Clyybber) [reply compiler/renderer.nim:253]: Readded it, since it might be useful for future renderer improvements.

File: compiler/vmgen.nim (hunk 7 above)
Comment (Clyybber) [reply compiler/vmgen.nim:1363]: This is very old (2013) and was used in genCopy but is long unused and outdated.

**Comment by Clyybber** [comment]
@timotheecour 
* There used to be a point, but it gets checked style-insensitive both ways nowadays, so it doesn't matter.
* Sure, I can add the commit messages, they should give a good idea of whats been done
* I simply compiled the compiler and the tools with `hint[DefinedButNotUsedX]:on` and used git blame and grep to find out if they are really unused (and not just unused inside a `when false:` or commented out). I also noticed a few things that could be cleaned up while looking at the lexer/parser.

**Review by timotheecour — APPROVED** [review]

**Comment by Araq** [comment]
Assignments like  ' result = "" ' will become mandatory, we really want definite assignment analysis like C#, so don't remove them. And sorry but I also don't like that you removed my lookup tables. ``$enumValue`` should be a debug helper, and say eg.  ``tkIf`` and the token's display name is "if". Please don't conflate them, the ship has sailed and $ is for debugging.

**Comment by timotheecour** [comment]
> Assignments like ' result = "" ' will become mandatory

until something like that happens, it doesn't make sense to have assignments in a some parts of the code code and not in others parts (other parts of stdlib, or, more importantly, all the third party code ever written that relies on `result` being assigned. In any case for something like that to happen, there would need to be an RFC + a tool to help with such edits (otherwise it's gonna be really painful), so that tool would be used to fix those anyways.

> I also don't like that you removed my lookup tables. $enumValue should be a debug helper

tkIf => if is more than just for debugging, it's the canonical string used in parser, error messages etc. In fact we only use the canonical form for those enums (`"if"`), not the raw form (`"tkIf"`); at the very least the canonical form is the one that's used most of the time. This PR makes things DRY and simpler to read/write/maintain.

What we can add (and I can volunteer an implementation) is a way to get the native string for an enum:
```nim
type Foo = enum
  f1 = "myF1"
  f2 = "myF2" 

import std/typetraits
const f = f1
static: doAssert $f == "myF1"
static: doAssert f.nativeStr == "f1"
```

**Comment by Araq** [comment]
> until something like that happens, it doesn't make sense to have assignments in a some parts of the code code and not in others parts (other parts of stdlib, or, more importantly, all the third party code ever written that relies on result being assigned.

Well the warning already exists and I'll enable it for the compiler. And I've already begun to patch the stdlib. Since the warning already exists, it doesn't need an RFC process for the time being, the compiler's codebase can diverge from the historical default settings.

I don't like the ``enumValue = "string repr"`` pattern and you won't convince me, differences in taste exist. So let me please have it my way.

**Comment by timotheecour** [comment]
> I don't like the enumValue = "string repr" pattern and you won't convince me, differences in taste exist. So let me please have it my way.

starting from first principles, a DRY solution is better. It doesn't have to be `enumValue = "string repr"` but can be a library solution. Can we then reconsider https://github.com/nim-lang/Nim/pull/14008 (with whatever adjustments needed) to be part of stdlib instead of fusion, it's a very common pattern (including for compiler code) with enums and generalizes  `enumValue = "string repr"` to arbitrary values/tuples to include other metadata: I've already shown several use cases in that PR and have more in mind. It solves the DRY aspect and also doesn't bake in strings with `enumValue = "string repr"`.

> enums with holes are getting fixed.

"enum with holes" in the title was just contextual, it's really not specific to enum with holes if you follow that PR; it's really about providing "enum map" with the simplest possible syntax and minimal overhead.

**Comment by Araq** [comment]
> Can we then reconsider #14008 

Yeah, that's a nice compromise. Bring it on please!

**Comment by timotheecour** [comment]
done, I suggest putting this PR on temporary hold pending #14008

**Comment by stisa** [comment]
I think the module [compiler/forloops](https://github.com/nim-lang/Nim/blob/devel/compiler/forloops.nim) is also dead code, or at least I can't find it used anywhere, so you may want to delete it in this PR?

File: compiler/btrees.nim [hunk 8]
```diff
@@ -131,12 +131,9 @@ proc toString[Key, Val](h: Node[Key, Val], indent: string; result: var string) =
       if j > 0: result.add(indent & "(" & $h.keys[j] & ")\n")
       toString(h.links[j], indent & "   ", result)
 
-proc `$`[Key, Val](b: BTree[Key, Val]): string =
-  result = ""
-  toString(b.root, "", result)
+proc `$`[Key, Val](b: BTree[Key, Val]): string = toString(b.root, "", result)
```
Comment (Araq) [inline compiler/btrees.nim]: Please revert, we agreed on leaving ``result = ""`` statements.

File: compiler/commands.nim [hunk 9]
```diff
@@ -53,11 +53,9 @@ const
       "Copyright (c) 2006-" & copyrightYear & " by Andreas Rumpf\n"
 
 proc genFeatureDesc[T: enum](t: typedesc[T]): string {.compileTime.} =
-  var x = ""
-  for f in low(T)..high(T):
-    if x.len > 0: x.add "|"
-    x.add $f
-  x
+  for f in T:
```
Comment (Araq) [inline compiler/commands.nim:57]: add ``result = ""``

**Pull request merged**